tracing/src/main/java/com/palantir/tracing/TraceLocal.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public final class TraceLocal<T> {
+
+    private static final SafeLogger log = SafeLoggerFactory.get(TraceLocal.class);
+
+    private final Supplier<T> initialValue;
+
+    /**
+     * Creates a trace local variable.
+     *
+     * When not already set, and this variable is accessed in a trace with the {@link #get()} method, the supplier
+     * will be invoked to supply a value (which will then be stored for future accesses).
+     *
+     * The supplier is thus normally invoked once per trace, but may be invoked again in case of subsequent
+     * invocations of {@link #remove()} followed by get.
+     */
+    public TraceLocal(Supplier<T> initialValue) {
```

Comment:
Should we require a supplier in all cases? is `T` nullable?

Reply:
I think we should make the ctor private, and require a builder for TraceLocal initialization. That way we can optionally allow an initial-value-provider (so get operations don't add data to the internal map unnecessarily) and add the listener interface easily in a follow-up.

Reply:
builder seems overkill for now, but I've made ctor private/added static factory methods